### ISSUE TITLE
[Ray Train] Ray Train running slow when multiple workers executed

### ISSUE THREAD

**Issue by Anando304** [ray-project/ray#27107]
## Context
I am trying to run a TensorFlow Model on a Ray Cluster running on K8's.

I am using [Ray Train](https://docs.ray.io/en/latest/train/train.html) to try to reduce the Training time. 

The dataset is a public Google dataset found here:
- `gs://cloud-training-demos/taxifare/ch4/taxi_preproc/`
- You can download locally via: `gsutil cp gs://cloud-training-demos/taxifare/ch4/taxi_preproc .`

I am running the Ray Cluster on Kubernettes without Helm:
- [cluster_crd.yaml](https://drive.google.com/file/d/1P1YRxwHAG8GOXWUmXw27OhZxCYJojU2t/view?usp=sharing)
- [operator_cluster_scoped.yaml](https://drive.google.com/file/d/1rXaBWM09gg1wfFVdxcEN7lfpNLdeXU2E/view?usp=sharing)
- [example_cluster.yaml](https://drive.google.com/file/d/1r0yEbi_TeqvK7z2lJ5w3dsFp3iR778B_/view?usp=sharing)

I am using the following [Jupyter Notebook](https://drive.google.com/file/d/1ZuumaXQgkd2x9qmIH1u2iUMRu7nKTPiz/view?usp=sharing) for my model.

Steps to reproduce can be found in the __Reproduction script__ section

## Problem
However, it seems as though the time it takes to compute with more than a single worker is still somewhat slow as compared to a single worker. 

From the Ray dashboard shown below, it seems that one worker is being heavily used most of the time while the other is not? 🤔 

![screenshot](https://user-images.githubusercontent.com/45672939/181350767-0a24b1ca-5ac1-497a-8905-e6a3161266f4.png)

**Question**
The behaviour of the notebook results make it seem as though it is running in sequential mode rather than distributed.
Is there a way to determine if something is running in parallel correctly? 

The dataset is roughly ~500MB, so its large enough that it doesn't finish training instantly in non-distributed mode, but its strange how much slower it is in distributed mode.

I've outlined the steps in the **Reproduction script** section for reproducing the environment

## System Specs:
Python 3.7.12
Ray 0.13.0
TensorFlow 2.9.1

**Ray Cluster Resources:**
- 'object_store_memory': 7585179647.0,
 - 'node:7.255.241.148': 1.0,
 - 'CPU': 12.0,
 - 'node:7.255.240.22': 1.0,
-  'memory': 18038862642.0,
-  'example-resource-b': 2.0,
-  'node:7.255.241.149': 1.0,
- 'example-resource-a': 2.0

## 1.0 Deploy Ray Cluster on K8's without Helm

Necessary Files:
- [cluster_crd.yaml](https://drive.google.com/file/d/1P1YRxwHAG8GOXWUmXw27OhZxCYJojU2t/view?usp=sharing)
- [operator_cluster_scoped.yaml](https://drive.google.com/file/d/1rXaBWM09gg1wfFVdxcEN7lfpNLdeXU2E/view?usp=sharing)
- [example_cluster.yaml](https://drive.google.com/file/d/1r0yEbi_TeqvK7z2lJ5w3dsFp3iR778B_/view?usp=sharing)

Commands (to be executed in order):
- `kubectl create -n ray -f ./src/python/twitter/ml/ray/cluster_config/cluster_crd.yaml`
- `kubectl create -n ray -f ./src/python/twitter/ml/ray/cluster_config/operator_cluster_scoped.yaml`
- `kubectl create -n <NAMESPACE> -f ./src/python/twitter/ml/ray/cluster_config/example_cluster.yaml`

## 2.0 Run Jupyter Notebook Model

Run the following [Jupyter Notebook](https://drive.google.com/file/d/1ZuumaXQgkd2x9qmIH1u2iUMRu7nKTPiz/view?usp=sharing) for my model.

**Section 6** in the notebook describing Distributed Training is where it seems to be slow. **Section 5** describes single worker training which seems to be faster.

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by xwjiang2010** [comment]
Thanks for filing this issue!
Seeing 
```
trainer = Trainer(backend="tensorflow", num_workers=2)

# For GPU Training, set `use_gpu` to True.
# trainer = Trainer(backend="tensorflow", num_workers=4, use_gpu=True)

trainer.start()
results = trainer.run(train_func_distributed)
trainer.shutdown()
```

@Anando304 It seems you were setting `num_workers=2` and from the screenshot, there were indeed two "workers" (PID: 2330 and 2331).
"Worker" in this context refers to a Ray actor (not a machine).

**Comment by Anando304** [comment]
Hey @xwjiang2010 !

Thanks for the response. That's good to know. However when I run the notebook, it's printing the epochs as if its running in non-distributed mode. Is it normally supposed to be this slow compared to the non-distributed mode?

**Comment by xwjiang2010** [comment]
@Anando304 
I tried your workload and indeed the distributed computing is not bringing benefit to the table.
The reason is following: 
For non-distributed one, the CPU usage is already around 100% (tf by default will use all the cores on the machine). For distributed one, the default set up is 1 CPU per worker and default placement for ray train workers is "PACKED" (on the same machine). As a result, the workload is still limited by the CPU capacity on single machine.

If you could try settings `TRAIN_ENABLE_WORKER_SPREAD_ENV` env variable and set `resources_per_worker` to use e.g. 4 CPUs (so that they can be spread across multiple machines), you can see an improvement with distributed training.

**Comment by Anando304** [comment]
Hi @xwjiang2010,

Thanks for the feedback! Your suggestions does improve the distributed training as shown in this notebook [here](https://drive.google.com/file/d/1IJvn7yiE5pP-tw8oznNGNvSgcu2Y5pjH/view?usp=sharing).

However, it is still roughly 10 seconds slower than non-distributed mode. I utilized 2 workers and 10 CPUs each. I guess overhead of distributed computing may be one reason among many, but still good to see better results compared to before. This falls inline with the performance results provided by the Ray benchmarks as shown [here](https://docs.ray.io/en/master/ray-air/benchmarks.html#tensorflow-training-parity).

I suppose Ray Train is good in specific cases only, but atleast the numbers line up with what Ray provided on the website. Thanks for your support.

**Comment by cszhu** [comment]
This P2 issue has seen no activity in the past 2 years. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please comment and remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
